cudaPackages: name and consolidate use of fixups pattern

ConnorBaker opened the pull request:
Consolidate fixups in a single `fixups` directory in preparation for merging more improvements from https://github.com/connorbaker/cuda-packages.

- `fixups`: Each file or directory (excluding `default.nix`) should contain a `callPackage`-able expression to be provided to the `overrideAttrs` attribute of a package produced by the generic manifest builder.
- Fixups are now applied by the manifest builder.
- Fixups have been updated where necessary to avoid possible missing attribute errors caused by use of attributes which did not exist in `prevAttrs`.
- Fixups used by the multiplexed builder (e.g., cuDNN) optionally take redist packages as arguments (defaulting to `null`) to handle the case where the multiplexed package is built against a version of CUDA predating redistributable packages (i.e., pre-11.4).
- The manifest builder now requires `redistName`.
- The manifest builder now adds `redistribRelease` and `featureRelease` to `passthru` for use by fixups. For example, TensorRT's fixup uses these attributes to select an appropriate version of cuDNN.
- `cuda-library-samples` now optionally takes redist packages as arguments to handle the case where a package is unavailable when building for a CUDA predating redistributable packages (i.e., pre-11.4).
- The extensions in `pkgs/top-level/cuda-packages.nix` are now `import`-ed rather than `callPackage`'d.

> [!NOTE]
>
> This PR doesn't aim to fix `cuda-library-samples`. It's still just as broken as it was before this PR.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

ConnorBaker committed:
cudaPackages: name and consolidate use of fixups pattern

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker commented:
Rebased and force-pushed on #404686.

SomeoneSerge commented on pkgs/development/cuda-modules/cuda-library-samples/generic.nix:
```diff
@@ -125,22 +121,36 @@ in
+        '';
+
+      installPhase =
+        prevAttrs.installPhase or ""
+        + ''
+          runHook preInstall
```
Note:

- `preInstall` should probably run before `prevAttrs.installPhase`
- `prevAttrs.installPhase` might already be running the hooks

SomeoneSerge replied:
I'd suggest you extend one of the (pre/post) hooks instead of the phase

ConnorBaker replied:
Switched to `postInstall` in the latest.

SomeoneSerge commented on pkgs/development/cuda-modules/fixups/cuda/cuda_nvcc.nix:
```diff
@@ -0,0 +1,77 @@
+  # Though it might seem odd or counter-intuitive to add the setup hook to `propagatedBuildInputs` instead of
+  # `propagatedNativeBuildInputs`, it is necessary! If you move the setup hook from `propagatedBuildInputs` to
+  # `propagatedNativeBuildInputs`, it stops being propagated to downstream packages during their build because
+  # setup hooks in `propagatedNativeBuildInputs` are not designed to affect the runtime or build environment of
+  # dependencies; they are only meant to affect the build environment of the package that directly includes them.
```
I thought there used to be a different message here (talking about offsets), and I'm kind of confused by the new one. "only meant to affect the build environment of the package that directly includes them"?

ConnorBaker replied:
It's the same one: https://github.com/NixOS/nixpkgs/blob/8b19a28ab7a307543536e0253aeacb2941672419/pkgs/development/cuda-modules/cuda/overrides.nix#L252-L257

Replaced in the latest.

ConnorBaker replied:
It's the same one, but I've replaced it with something more succinct in the latest.

ConnorBaker replied:
That one I'm not sure about :(

SomeoneSerge replied:
TLDR: this part of the comment is erroneous too, so I'd leave it out while we're at it.

Btw, I think I get it now: `buildInputs` means "same (host) platform as the current package", so relative to us their offsets are `(0, 0)`, and `nativeBuildInputs` are shifted by `(-1, -1)`. My original code was adding `(1, 0)` instead of the current `(0, 0)`

SomeoneSerge commented on pkgs/development/cuda-modules/fixups/cuda/cuda_nvcc.nix:
```diff
@@ -0,0 +1,77 @@
+      nvvmReplace = lib.optionalString (cudaOlder "12.5") ''
+        --replace-fail \
+          '$(TOP)/$(_NVVM_BRANCH_)' \
+          "''${!outputBin}/nvvm" \
```
H'm, has this forward-slash always been here? Seems dangerous, unless we never put `nvvmReplace` last

ConnorBaker replied:
Made a little less dangerous-looking in the latest.

SomeoneSerge commented on pkgs/development/cuda-modules/fixups/default.nix:
```diff
@@ -0,0 +1,40 @@
+      }
+    ) { } (readDir fixupFnDir);
+in
+foldlAttrs (
+  acc: fileName: fileType:
+  acc
+  // optionalAttrs (fileType == "directory") {
+    ${fileName} = mkFixups (./. + "/${fileName}");
+  }
+) { } (readDir ./.)
```
Is there much value to putting `cuda/cuda_nvcc.nix` and `cudnn/cudnn.nix` into per-distribution directories? The final `cudaPackages` are flat. We could as well start making everything flat, and this then becomes another `packagesFromDirectoryRecursive`

ConnorBaker replied:
I had considered that, but I like the grouping a lot because tells us where a package comes from and it mirrors the structure of `manifests`: https://github.com/ConnorBaker/cuda-packages/tree/main/pkgs/development/cuda-modules/manifests.

SomeoneSerge replied:
Now that "fixup"s are `callPackage`d, which `override` function do packages inherit? The one from `extension`s, or the one from `fixups`

ConnorBaker replied:
They inherit the one from the generic manifest builder:

```console
nix eval --impure --json .#cudaPackages.cuda_nvcc.override.__functionArgs
```

```json
{
  "autoAddCudaCompatRunpath": false,
  "autoAddDriverRunpath": false,
  "autoPatchelfHook": false,
  "backendStdenv": false,
  "callPackage": false,
  "cudaMajorMinorVersion": false,
  "featureRelease": false,
  "fetchurl": false,
  "fixups": false,
  "flags": false,
  "lib": false,
  "libPath": true,
  "markForCudatoolkitRootHook": false,
  "pname": false,
  "redistName": false,
  "redistribRelease": false,
  "stdenv": false
}
```

```console
nix eval --impure --json .#cudaPackages.cudnn.override.__functionArgs
```

```json
{
  "autoAddCudaCompatRunpath": false,
  "autoAddDriverRunpath": false,
  "autoPatchelfHook": false,
  "backendStdenv": false,
  "callPackage": false,
  "cudaMajorMinorVersion": false,
  "featureRelease": false,
  "fetchurl": false,
  "fixups": false,
  "flags": false,
  "lib": false,
  "libPath": true,
  "markForCudatoolkitRootHook": false,
  "pname": false,
  "redistName": false,
  "redistribRelease": false,
  "stdenv": false
}
```

SomeoneSerge replied:
Ah, I completely missed that you're building a (well, flat) tree of Path's, not of results of `callPackage`. This bit is good then: I think it's `packagesFromDirectoryRecursive` and TVL's `readTree` that need to be (abandoned and) rewritten to separate the capture of the tree from the auto-calling.

However, I'm still skeptical about reflecting the layout of the manifests: I'd just make this a single flat `pkgs/development/cuda-modules/overrides`. This would clearly suggest: if you want to modify the behaviour of `cudaPackages.${name}`, add/edit `overrides/${name}.nix`

> These fixups need to be callPackage'd by the final CUDA package set scope.

Is that a problem?

ConnorBaker replied:
> However, I'm still skeptical about reflecting the layout of the manifests: I'd just make this a single flat pkgs/development/cuda-modules/overrides. This would clearly suggest: if you want to modify the behaviour of cudaPackages.${name}, add/edit overrides/${name}.nix

I'll make `fixups` flat.

I'm hesitant to call them overrides given the term seems a bit overloaded. I want to make clear that a `fixup` is an expression which is `callPackage`'d and then provided to a package's `overrideAttrs` attribute.

See https://github.com/ConnorBaker/cuda-packages/blob/main/doc/language-frameworks/cuda.section.md#configuring-cuda-package-sets-cuda-configuring-cuda-package-sets and https://github.com/ConnorBaker/cuda-packages/blob/main/doc/language-frameworks/cuda.section.md#creating-cuda-package-sets-cuda-creating-cuda-package-sets for what modifying or creating the package set would look like. I need to make it easy to do so out-of-tree, thus passing the `fixups` and `manifests` as `callPackage` arguments in `pkgs/top-level/cuda-packages.nix` vs. reading from the directory in the construction of each package set.

> Is that a problem?

Ah, just wanted to make it clear that they must use the scope's `callPackage`.

SomeoneSerge replied:
![image](https://github.com/user-attachments/assets/0ba1a241-0b23-4226-820f-c14cfd947ab8)

Maintaining "overrides" for code-generated packages is a standard practice, and all of these overrides are functions of package sets, not just `finalAttrs: prevAttrs:`. The difference is you split `overrides.nix` into multiple files. I sort of see the point, that the files container "override constructors" which produce the actual "override functions", but what's heaver on the reader's brain, a new term, or the slight abuse of a very familiar term with a bunch of ready examples in the directory?

SomeoneSerge replied:
(obviously depends on the reader's background though)

ConnorBaker replied:
Good point! I'll switch over to `overrides` then.

SomeoneSerge replied:
> They inherit the one from the generic manifest builder:

Good, I guess there's no change then. However, for future, we should remember ourselves whether `override` was made to commute `overrideAttrs` yet (I think it wasn't, which means that, unless we keep applying them inside the generic builder, calling `override` can erase the fixups (Ha! Just now it was actually very convenient to have a word for the function that goes into `overrideAttrs` as opposed to into `override`; not quite the meaning you had implied originally, but it worked))

ConnorBaker replied:
Good point! I added a note in the manifest builder.

SomeoneSerge commented on pkgs/top-level/cuda-packages.nix:
```diff
@@ -127,30 +143,53 @@ let
       })
-      (callPackage ../development/cuda-modules/cuda-samples/extension.nix { inherit cudaVersion; })
-      (callPackage ../development/cuda-modules/cuda-library-samples/extension.nix { })
+      (import ../development/cuda-modules/cuda-samples/extension.nix { inherit cudaVersion lib stdenv; })
+      (import ../development/cuda-modules/cuda-library-samples/extension.nix { inherit lib stdenv; })
```
Why these?

ConnorBaker replied:
Why convert them to `import`? Because we benefit directly from import caching, rather than hiding `import` inside of a `callPackage`.

SomeoneSerge replied:
Ah so the difference is it's easier to prove the result doesn't depend on `self`? I wonder if there's any source other than C++ sources that actually clearly lays out our memoization rules...

ConnorBaker replied:
Nix maintains a cache for parsed files; see https://github.com/NixOS/nix/blob/1e822bd4149a8bce1da81ee2ad9404986b07914c/src/libexpr/primops.cc#L282 and https://github.com/NixOS/nix/blob/1e822bd4149a8bce1da81ee2ad9404986b07914c/src/libexpr/eval.cc#L1122

SomeoneSerge replied:
But that doesn't necessarily mean there's a difference with `callPackage`

ConnorBaker replied:
Only that we can lift it up to the top, drop `callPackage`'s (small) overhead, and don't need to worry about them being `override`-able.

SomeoneSerge replied:
> can lift it up to the top

This, I didn't know how to phrase it, but this is what I meant by "prove no dependency on `self`"

SomeoneSerge commented on pkgs/development/cuda-modules/cuda/overrides.nix:
Going to go over the diff one more time, but I feel like merging as soon as ~~ofborg~~ eval is finished

ConnorBaker replied:
Okay! The latest force-push changed the `maybeFixup` value in the manifest builder to `maybeOverride` and added a note about what you pointed out in https://github.com/NixOS/nixpkgs/pull/404973#discussion_r2078558057.

ConnorBaker replied:
Force-pushed with the fix for `mkVersionedPackageName` which was causing the missing attribute error for `cupy`. (Gotta string interpolation typos.)

SomeoneSerge commented on pkgs/development/cuda-modules/cuda/overrides.nix:
> ```
> error: attribute 'cudnn_8_9' missing
> cudnn = cu-fi.cudnn_8_9;
> ```

Honestly, this should have been `cu-pre.cudnn.override ...`, or, in fact, `cudaPackages.override { ....(different metadata database) }`

ConnorBaker replied:
I like the second form best ;)

SomeoneSerge commented on pkgs/development/cuda-modules/overrides/default.nix:
```diff
@@ -0,0 +1,9 @@
+    value = import (./. + "/${name}");
+  })
+  # Make sure to remove this file (default.nix) from the list of files/directories.
+  (lib.removeAttrs (builtins.readDir ./.) [ "default.nix" ])
```
Now we're pulling nits, but I suspect there's a complexity difference between this and `concatMap` xDDD

SomeoneSerge commented on pkgs/development/cuda-modules/generic-builders/manifest.nix:
```diff
@@ -40,6 +41,9 @@ let
   inherit (stdenv) hostPlatform;
 
+  # Override function for the package.
+  maybeOverride = overrides.${pname} or null;
```
Now this is kind of different from `overrides.nix` in the sense that we have an attribute `cudnn_8_9`, and you'd expect that adding `overrides/cudnn_8_9.nix` would affect it but it won't.

However, exactly because the "cudnn.nix" "fixup" is applied to all instances of "cudnn" coming from manifests, the semantics here is different from "just apply overrides to attributes in the very last step of composition"

SomeoneSerge replied:
Alright second attempt at interpretation: "fixups" are "early overrides", in the sense that they are applied before we create `.override`

SomeoneSerge commented on pkgs/top-level/cuda-packages.nix:
```diff
@@ -20,13 +20,35 @@
 # extensions need to compute the attribute names, without relying on `final`.
 #
 # I've (@connorbaker) attempted to do that, though I'm unsure of how this will interact with overrides.
+let
+  # Import caching
+  lib = import ../../lib;
+  overrides = import ../development/cuda-modules/overrides { inherit lib; };
+  gpus = import ../development/cuda-modules/gpus.nix;
+  nvccCompatibilities = import ../development/cuda-modules/nvcc-compatibilities.nix;
+
+  # Extensions
+  aliasExtension = import ../development/cuda-modules/aliases.nix;
+
+  # Extension-building functions
+  mkMultiplexExtension = import ../development/cuda-modules/generic-builders/multiplex.nix;
+  mkCudaExtension = import ../development/cuda-modules/cuda/extension.nix;
+  mkCudaSamplesExtension = import ../development/cuda-modules/cuda-samples/extension.nix;
+  mkCudaLibrarySamplesExtension = import ../development/cuda-modules/cuda-library-samples/extension.nix;
+  mkCuTensorExtension = import ../development/cuda-modules/cutensor/extension.nix;
+  mkCuSparseLtExtension = import ../development/cuda-modules/cusparselt/extension.nix;
```
Do you know for sure that moving things up to the let-in makes a difference? I mean if you `import ./foo.nix a b c d` and all of `a b c d` come from the outermost scope (independent of `self`), can't Nix figure out that it doesn't need to re-parse/re-import stuff?

SomeoneSerge replied:
Honestly, unless there's a minimal example that was already benchmarked and that demonstrates the issue, I'd say let's go with for the cleanest looks and optimize performance later

ConnorBaker replied:
Per https://github.com/NixOS/nixpkgs/pull/404973#discussion_r2078465508, let-in floating only reduces the number of times `import` is called. Generally (I think), it also allows for prevent duplicate thunks, since as thunks are forced in the values in the top scope they stay in scope. In contrast, in the lower position, function applications (like those in `../development/cuda-modules/overrides/default.nix`) would need to be re-evaluated.

SomeoneSerge replied:
> In contrast, in the lower position, function applications

I see how that could happen but this also sounds sort of silly, because why can't Nix "lift" things automatically when they don't depend on anything in the inner scope

ConnorBaker replied:
Nix definitely could, it doesn't currently because that sort of functionality falls under the purview of an optimizing interpreter ;)

ConnorBaker commented:
Okay, force-pushed to change `overrides` back to `fixups` (and making a note about how there's no "fixup" for versioned packages since they only apply to the base package, by pname) and floated the imports in `pkgs/top-level/cuda-packages.nix` back down and inline.

SomeoneSerge commented on pkgs/development/cuda-modules/cusparselt/extension.nix:
Note for future: sort of inconvenient that we have pname = default attribute name = file name throughout the package set, but then cusparse_lt is in cusparselt/

ConnorBaker replied:
I plan to change it >:)

SomeoneSerge commented on pkgs/top-level/cuda-packages.nix:
```diff
@@ -38,17 +38,45 @@ let
-  gpus = builtins.import ../development/cuda-modules/gpus.nix;
-  nvccCompatibilities = builtins.import ../development/cuda-modules/nvcc-compatibilities.nix;
-  flags = callPackage ../development/cuda-modules/flags.nix { inherit cudaVersion gpus; };
+  # As values created outside of the package set fixed-point, the presence of attributes (and their names) can rely on
+  # these.
+  # By that same proprety, changes to these values will not be reflected in the package set, since they exist outside
+  # of the fixed-point.
+  fixups = import ../development/cuda-modules/fixups { inherit lib; };
+  gpus = import ../development/cuda-modules/gpus.nix;
```
> `# these`

I suppose this is a reminder that "Nix is strict in namesets" which is why "it's significant fixups are independent of the fix-point"

> will not be reflected

These values don't depend on the fix-point, but the fix-point does depend on these values.

SomeoneSerge commented on pkgs/development/cuda-modules/generic-builders/manifest.nix:
```diff
@@ -321,7 +329,26 @@ backendStdenv.mkDerivation (finalAttrs: {
+    # Perform package-specific fixup, if it exists.
+    # NOTE: If we do not apply the fixup function here, the `override` attribute added by `callPackage` could return a
+    # version of this derivation without the fixup having been applied.
+    if maybeFixup != null then callPackage maybeFixup { } else { }
```
could unwrap `maybeFixup` earlier in the let-in

SomeoneSerge commented on pkgs/development/cuda-modules/generic-builders/multiplex.nix:
For a moment I thought it was a side effect of https://github.com/NixOS/nixpkgs/pull/404686/files#diff-ed90252d158338eb264d6c0ef99282d6b5a3d37f1ed84f63d2fc7a208d12a10c, but whatever the reason `cudnn-frontend` was already taking `cudaPackages` as the argument when it was introduced, instead of taking `cudnn ? null`

SomeoneSerge replied:
It is partially a result of https://github.com/NixOS/nixpkgs/pull/404686 in the sense that now we add elements of `packages/` to the package set unconditionally (which is more consistent than filtering imo). I'ma just mark it unsupported

SomeoneSerge committed:
cudaPackages_11_0.cudnn-frontend: mark unsupported, prevent eval error

With https://github.com/NixOS/nixpkgs/pull/404686 we add packages/ unconditionally

ConnorBaker commented:
Good fixes! Thank you :)

ConnorBaker commented:
I'm happy with the changes, so feel free to merge when ready @SomeoneSerge!
